README


This archive contains the distribution Giovanni,
version 1.11:

  a Perl based deployment system

This software is copyright (c) 2014 by ideegeo Group Limited.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


This README file was generated by Dist::Zilla::Plugin::Readme v5.008.



Changes
==================================================
Changes from 2014-10-09 00:00:00 +0000 to present.
==================================================

-----------------------------------------
version 1.11 at 2015-10-09 04:43:28 +0000
-----------------------------------------

  Change: a30d7e75bc0d597d1a7ee6622318d813b57f80a1
  Author: Lenz Gschwendtner <lenz@springtimesoft.com>
  Date : 2015-10-09 17:42:59 +0000

    load LWP::Protocol::https explicitly so it gets pulled in as a
    dependency 

  Change: 7dcceaf30b1130286bd4df36e9b72e597a17103c
  Author: Lenz Gschwendtner <lenz@springtimesoft.com>
  Date : 2014-10-31 12:25:24 +0000

    v1.10 

-----------------------------------------
version 1.10 at 2014-10-31 04:24:05 +0000
-----------------------------------------

  Change: 2dc4aeaa95965b4502bcb8501dc49fd13dde0585
  Author: Lenz Gschwendtner <lenz@springtimesoft.com>
  Date : 2014-10-31 12:24:05 +0000

    new README 

  Change: cc107020801c918b996085794e4344d467ddefde
  Author: Lenz Gschwendtner <lenz@springtimesoft.com>
  Date : 2014-10-31 12:23:38 +0000

    fix version number 

  Change: 2b02b07eb62c597115ba561db9c4f9ea904dae93
  Author: Lenz Gschwendtner <lenz@springtimesoft.com>
  Date : 2014-10-31 12:19:55 +0000

    drop jabber support as it requires LibIDN which is a pain to build 

  Change: f41d66271023dec0694d9f5f1f320c2ced60274d
  Author: Lenz Gschwendtner <lenz@springtimesoft.com>
  Date : 2014-04-15 18:45:16 +0000

    dzillify project 

  Change: 4a20030af3c5cfdb8921037c09b2c2216357bec1
  Author: Lenz Gschwendtner <lenz@springtimesoft.com>
  Date : 2014-04-15 17:50:33 +0000

    added a reload_pased stage 

  Change: 07fddd7fad6493deddf8f3919ab1601317a914be
  Author: Lenz Gschwendtner <lenz@springtimesoft.com>
  Date : 2013-02-28 13:47:32 +0000

    Checking in changes prior to tagging of version 1.8.

    Changelog diff is:

    diff --git a/Changes b/Changes index b215c7e..c418dd7 100644 ---
    a/Changes +++ b/Changes @@ -1,5 +1,8 @@

    Revision history for Giovanni

    +1.8 2013-02-28T13:46:23 + added a username to the deploy message +

    1.7 2013-02-28T13:04:44

    - removed XMPP dependency as it is horrible to install

    - added LWP::UserAgent for the HipChat plugin 

  Change: c6c3d66a18c97775ba2b72c4153e58e7050bfc02
  Author: Lenz Gschwendtner <lenz@springtimesoft.com>
  Date : 2013-02-28 13:05:51 +0000

    Checking in changes prior to tagging of version 1.7.

    Changelog diff is:

    diff --git a/Changes b/Changes index 0e87176..b215c7e 100644 ---
    a/Changes +++ b/Changes @@ -1,5 +1,9 @@

    Revision history for Giovanni

    +1.7 2013-02-28T13:04:44 + - removed XMPP dependency as it is
    horrible to install + - added LWP::UserAgent for the HipChat plugin +

    1.6 2012-08-09T16:05:19

    added restart as a separate function 

  Change: c06fc8f5045c87b66c35a3313a4f68abbfd2c9df
  Author: Lenz Gschwendtner <lenz@springtimesoft.com>
  Date : 2012-08-09 16:05:39 +0000

    Checking in changes prior to tagging of version 1.6.

    Changelog diff is:

    diff --git a/Changes b/Changes index 779355a..0e87176 100644 ---
    a/Changes +++ b/Changes @@ -1,5 +1,8 @@

    Revision history for Giovanni

    +1.6 2012-08-09T16:05:19 + added restart as a separate function +

    1.5 2012-08-09T15:42:29

    added support for HipChat 

  Change: 80b0ef2c597b638eb36090e064b4088c7aa23013
  Author: Lenz Gschwendtner <lenz@springtimesoft.com>
  Date : 2012-08-09 15:44:25 +0000

    Checking in changes prior to tagging of version 1.5.

    Changelog diff is:

    diff --git a/Changes b/Changes index eff130c..779355a 100644 ---
    a/Changes +++ b/Changes @@ -1,5 +1,8 @@

    Revision history for Giovanni

    +1.5 2012-08-09T15:42:29 + added support for HipChat +

    1.4 2012-07-13T12:18:32

    implemented transparent post_rollout script stage [lev] 

  Change: 29949ef5b0aca64cd59f1575a0a80b8935ed48c8
  Author: Lenz Gschwendtner <lenz@springtimesoft.com>
  Date : 2012-08-09 15:41:39 +0000

    added hipchat support 

  Change: 4945f4feaf513bcf4a36eaac638c0bb519baf4ce
  Author: Lenz Gschwendtner <lenz@springtimesoft.com>
  Date : 2012-08-09 15:27:54 +0000

    experimental hipchat support 

  Change: 064e1c3c1f4552931941b0502c991ab1f778fc39
  Author: Lenz Gschwendtner <lenz@springtimesoft.com>
  Date : 2012-07-16 16:46:32 +0000

    fixed rollback linking 

  Change: 74e17fa47baec3ef43156edf7e4b47c50202850c
  Author: Lenz Gschwendtner <lenz@springtimesoft.com>
  Date : 2012-07-16 16:34:30 +0000

    Checking in changes prior to tagging of version 1.4.

    Changelog diff is: 

  Change: 6630e6bc193d6e0da1e45c05051681243eb6d60b
  Author: Tobias Kirschstein <lev@ideegeo.com>
  Date : 2012-07-13 13:27:14 +0000

    implemented transparent post_rollout script stage 

  Change: 6eaf59762d3b1b26d5d11d90c9607133ce572369
  Author: Lenz Gschwendtner <lenz@springtimesoft.com>
  Date : 2012-04-23 17:36:47 +0000

    Checking in changes prior to tagging of version 1.3.

    Changelog diff is:

    diff --git c/Changes w/Changes index 93643a7..609d730 100644 ---
    c/Changes +++ w/Changes @@ -1,5 +1,8 @@

    Revision history for Giovanni

    +1.3 2012-04-23T17:35:24 + fixed a jabber notify issue [lenz] +

    1.2 2012-02-15T13:34:19

    fixed bug with multiple adresses in the jabber notify plugin

    [lenz] 

  Change: e78c679e69a43734ab9cd3eb7fd0ca7c3c7c58a7
  Author: Lenz Gschwendtner <lenz@springtimesoft.com>
  Date : 2012-04-23 17:35:11 +0000

    added missing dependencies 

  Change: 13aa41da1e38f71bd8641c5eb5ad243b211d953a
  Author: Lenz Gschwendtner <lenz@springtimesoft.com>
  Date : 2012-04-23 17:33:37 +0000

    fixed jabber notify issue 

  Change: b7e9159cff68eb64c52709b9ea36f71952b195ee
  Author: Lenz Gschwendtner <lenz@springtimesoft.com>
  Date : 2012-02-15 13:34:51 +0000

    Checking in changes prior to tagging of version 1.2.

    Changelog diff is:

    diff --git c/Changes w/Changes index 5a204b4..93643a7 100644 ---

Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.008.
use strict;
use warnings;

use 5.010;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "a Perl based deployment system",
  "AUTHOR" => "Lenz Gschwendtner <mail\@norbu09.org>",
  "BUILD_REQUIRES" => {},
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "Giovanni",
  "EXE_FILES" => [
    "bin/gio"
  ],
  "LICENSE" => "perl",
  "NAME" => "Giovanni",
  "PREREQ_PM" => {
    "Config::GitLike" => 0,
    "Cwd" => 0,
    "Data::Dumper" => 0,
    "Expect" => 0,
    "Getopt::Long" => 0,
    "Git::Repository" => 0,
    "LWP::Protocol::https" => 0,
    "LWP::UserAgent" => 0,
    "Mouse" => 0,
    "Mouse::Role" => 0,
    "Mouse::Util" => 0,
    "Net::OpenSSH" => 0,
    "Pod::Usage" => 0,
    "Sys::Hostname" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "Test::More" => 0
  },
  "VERSION" => "1.11",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Config::GitLike" => 0,
  "Cwd" => 0,
  "Data::Dumper" => 0,
  "Expect" => 0,
  "Getopt::Long" => 0,
  "Git::Repository" => 0,
  "LWP::Protocol::https" => 0,
  "LWP::UserAgent" => 0,
  "Mouse" => 0,
  "Mouse::Role" => 0,
  "Mouse::Util" => 0,
  "Net::OpenSSH" => 0,
  "Pod::Usage" => 0,
  "Sys::Hostname" => 0,
  "Test::More" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);





lib/Giovanni/Stages.pm
package Giovanni::Stages;

use Mouse;
use Expect;
use Data::Dumper;

# Stages are defined here and expected to be overridden with plugins.
# the idea is to have different plugins that can extend the existing
# stages real easy. If this stages approach turns out to be too limited
# (ie 1000s of stages in one file, not a good look) we may need to
# rethink this approach.

sub update_cache {
    my ($self, $ssh) = @_;
    $self->log($ssh, "running update_cache task ...");
    return;
}

sub rollout {
    my ($self, $ssh) = @_;

    $self->log($ssh, "running rollout task ...");
    $self->config->{deploy_dir} = $self->config->{root};
    my $log = $ssh->capture("mkdir -p " . $self->config->{deploy_dir});
    $self->log($ssh, $log);
    $self->checkout($ssh);
    $self->post_rollout($ssh);

    return;
}

sub rollout_timestamped {
    my ($self, $ssh) = @_;

    $self->log($ssh, "running rollout_timestamped task ...");
    my $deploy_dir = join('/', $self->config->{root}, 'releases', time);
    my $current = join('/', $self->config->{root}, 'current');
    my $log = $ssh->capture("mkdir -p " . $deploy_dir);
    $self->config->{deploy_dir} = $deploy_dir;
    $self->checkout($ssh);
    $log .= $ssh->capture(
        "unlink " . $current . "; ln -s " . $deploy_dir . " " . $current);
    $self->log($ssh, $log);
    $self->post_rollout($ssh);

    return;
}

sub rollback_timestamped {
    my ($self, $ssh, $offset) = @_;

    $self->log($ssh, "running rollback task ...");
    my $deploy_dir = join('/', $self->config->{root}, 'releases');
    my $current    = join('/', $self->config->{root}, 'current');
    my @rels = $ssh->capture("ls -1 " . $deploy_dir);
    @rels = sort(@rels);
    my $link = $ssh->capture("ls -l " . $current . " | sed 's/^.*->\\s*//'");
    my @path = split(/\//, $link);
    my $current_rel = pop(@path);
    my (@past, @future);

    foreach my $rel (@rels) {
        chomp($rel);
        next unless $rel =~ m{^\w};
        if ($rel == $current_rel) {
            push(@future, $rel);
            next;
        }
        if (@future) {
            push(@future, $rel);
        }
        else {
            push(@past, $rel);
        }
    }
    $deploy_dir = join('/', $self->config->{root}, 'releases', pop(@past));
    my $log = $ssh->capture(
        "unlink " . $current . "; ln -s " . $deploy_dir . " " . $current);
    $self->log($ssh, $log);
    return;
}

sub rollback_scm {
    my ($self, $ssh, $offset) = @_;

    # load SCM plugin
    $self->load_plugin($self->scm);
    my $tag = $self->get_last_tag($offset);
    $self->log($ssh, "Rolling back to tag: $tag") if $self->is_debug;

    # TODO change checkout to accept an optional tag so we can reuse it
    # here to check out an old version.

    return;
}

sub restart {
    my ($self, $ssh) = @_;

    $self->log("running restart task ...");
    my $log = $ssh->capture($self->config->{init});
    $self->log($ssh, $log);
    return;
}

sub checkout {
    my ($self, $ssh) = @_;
    $self->log($ssh, "running checkout task ...");
    return;
}

sub cleanup_timestamped {
    my ($self, $ssh, $offset) = @_;

    $self->log($ssh, "running cleanup task ...");

    if ($self->config->{root} =~ m{^.*/\d+$}) {
        my @path = split(/\//, $self->config->{root});
        pop(@path);
        pop(@path);
        $self->config->{root} = join('/', @path);
    }

    my $deploy_dir = join('/', $self->config->{root}, 'releases');
    my $current    = join('/', $self->config->{root}, 'current');
    my @rels = $ssh->capture("ls -1 " . $deploy_dir);
    @rels = sort(@rels);
    my $link = $ssh->capture("ls -l " . $current . " | sed 's/^.*->\\s*//'");
    my @path = split(/\//, $link);
    my $current_rel = pop(@path);
    my (@past, @future);

    foreach my $rel (@rels) {
        chomp($rel);
        next unless $rel =~ m{^\w};
        if ($rel == $current_rel) {
            push(@future, $rel);
            next;
        }
        if (@future) {
            push(@future, $rel);
        }
        else {
            push(@past, $rel);
        }
    }
    $deploy_dir = join('/', $self->config->{root}, pop(@past));
    my $num = $self->config->{keep_versions} || 5;
    my $log;
    while ($#past > ($num)) {
        my $to_del = join('/', $self->config->{root}, 'releases', shift(@past));
        $self->log($ssh, "deleting $to_del");
        $log = $ssh->capture("rm -rf " . $to_del);
    }
    $self->log($ssh, $log);

    return;
}

sub restart_phased {
    my ($self, $ssh) = @_;

    $self->log($ssh, "running restart_phased task ...");
    unless ($ssh->test("sudo " . $self->config->{init} . " restart")) {
        $self->log(
            'restart failed: ' . $ssh->error . ' trying stop -> start instead');
        $ssh->test("sudo " . $self->config->{init} . " stop");

        # give time to exit
        sleep 2;
        $ssh->test("sudo " . $self->config->{init} . " start")
            or $self->error('restart failed: ' . $ssh->error);
        return;
    }

    # my $exp = Expect->init($pty);
    # $exp->interact();
    $self->log($ssh, 'restarted ' . $self->config->{init});

    return;
}

sub reload_phased {
    my ($self, $ssh) = @_;

    $self->log($ssh, "running reload_phased task ...");
    unless ($ssh->test("sudo " . $self->config->{reload})) {
        $self->log(
            'reload failed: ' . $ssh->error . ' trying stop -> start instead');
        $ssh->test("sudo " . $self->config->{init} . " stop");

        # give time to exit
        sleep 2;
        $ssh->test("sudo " . $self->config->{init} . " start")
            or $self->error('restart failed: ' . $ssh->error);
        return;
    }

    # my $exp = Expect->init($pty);
    # $exp->interact();
    $self->log($ssh, 'reloaded ' . $self->config->{reload});

    return;
}

sub send_notify {
    my ($self, $ssh) = @_;
    $self->log($ssh, "running send_notify task ...");
    return;
}

sub post_rollout {
    my ($self, $ssh) = @_;

    return
        unless (exists $self->config->{post_rollout}
        and defined $self->config->{post_rollout});

    $self->log($ssh,
              'running post_rollout script "'
            . $self->config->{post_rollout}
            . '" ...');
    unless (
        $ssh->test(
                  'cd '
                . $self->config->{deploy_dir} . ' && '
                . $self->config->{post_rollout}))
    {
        $self->error("post_rollout script failed: " . $ssh->error);
    }
    return;
}

__PACKAGE__->meta->make_immutable;

__END__

=pod

=encoding UTF-8

=head1 NAME

Giovanni::Stages

=head1 VERSION

version 1.11

=head1 AUTHOR

Lenz Gschwendtner <mail@norbu09.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by ideegeo Group Limited.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Giovanni/Plugins/Git.pm
package Giovanni::Plugins::Git;

use Mouse::Role;
use Git::Repository;

has 'git' => (
    is      => 'rw',
    isa     => 'Git::Repository',
    lazy    => 1,
    default => sub { Git::Repository->new(work_tree => $_[0]->repo) },
);

sub tag {
    my $self = shift;

    my $tag = 'v' . time;
    my $log =
        $self->git->run(tag => '-a', $tag, '-m', "tagging to $tag for rollout");
    $self->log('git', "Tag: [$tag] " . $log) if $self->is_debug;
    $log = $self->git->run('pull') unless $self->is_debug;
    $self->log('git', "Pull: " . $log) if $self->is_debug;
    $log = $self->git->run(push => 'origin', '--tags') unless $self->is_debug;
    $self->log('git', "Push: " . $log) if $self->is_debug;
    $self->version($tag);
    $self->log('git', $log);

    return $tag;
}

sub get_last_tag {
    my ($self, $n) = @_;

    $n = 1 unless $n;
    my @tags = $self->git->run('tag');
    $self->log('git', "Tags: " . join(', ', @tags) . $/) if $self->is_debug;

    return splice(@tags, $n - 1, $n);
}

around 'update_cache' => sub {
    my ($orig, $self, $ssh, $conf) = @_;

    my $log;
    my $cache_dir = $self->_get_cache_dir($conf);

    if ($cache_dir) {
        if ($ssh->test("[ -d " . $cache_dir . "/.git ]")) {
            $log .= $ssh->capture("cd $cache_dir && git pull");
            $self->log("running git pull ...");
        }
        else {
            $log = $ssh->capture("mkdir -p " . $cache_dir)
                unless $ssh->test("[ -d " . $cache_dir . " ]");
            $log .= $ssh->capture("cd ". $cache_dir
                . "&& git clone " . $self->config->{repo} . " .");
            $self->log("running git clone ...");
        }
    }
    $self->log($ssh, $log);

    return;
};

around 'checkout' => sub {
    my ($orig, $self, $ssh, $conf) = @_;

    my $log;
    my $cache_dir = $self->_get_cache_dir($conf);
    if ($self->config->{deploy_dir}) {
        $log .=
            $ssh->capture("cd "
                . $self->config->{deploy_dir}
                . " && git clone --depth 1 --no-hardlinks file://"
                . $cache_dir
                . " .");
    }
    $self->log($ssh, $log);
};

sub _get_cache_dir {
    my ($self, $conf) = @_;
    if($self->config->{cache}){
        my @parts = split(/\//, $self->config->{repo});
        my $git_dir = pop(@parts);
        return join('/', $self->config->{cache}, $git_dir);
    } else {
        return $self->config->{root};
    }
}

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Giovanni::Plugins::Git

=head1 VERSION

version 1.11

=head1 AUTHOR

Lenz Gschwendtner <mail@norbu09.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by ideegeo Group Limited.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Giovanni/Plugins/Jabber.pm
package Giovanni::Plugins::Jabber;

use Mouse::Role;
use AnyEvent;
use AnyEvent::XMPP::Client;
use AnyEvent::XMPP::Ext::Disco;
use AnyEvent::XMPP::Ext::Version;
use AnyEvent::XMPP::Ext::MUC;
use AnyEvent::XMPP::Ext::MUC::Message;
use Data::Dumper;

around 'send_notify' => sub {
    my ( $orig, $self, $ssh ) = @_;

    print "notify via jabber\n";
    my @tos = split(/\s*,\s*/, $self->config->{jabber_to});
    my $msg =
        'just ran a '
      . $self->config->{command} . ' for '
      . $self->config->{project} . ' on '
      . $ssh->get_host;
    my $j = AnyEvent->condvar;
    my $cl = AnyEvent::XMPP::Client->new();
    foreach my $to (@tos){
        my $cnf = 1 if ( $to =~ m{conference} );

        my ( $disco, $version, $muc, $muc_msg ) if $cnf;

        if ($cnf) {
            $disco   = AnyEvent::XMPP::Ext::Disco->new;
            $version = AnyEvent::XMPP::Ext::Version->new;
            $muc     = AnyEvent::XMPP::Ext::MUC->new( disco => $disco );
            $muc_msg = AnyEvent::XMPP::Ext::MUC::Message->new(
                type => 'groupchat',
                to   => $to,
                body => $msg
            );
            $cl->set_presence( undef, 'Giovanni is talking now.', 1 );
        }
        $cl->add_account( $self->config->{jabber_user}, $self->config->{jabber_pass} );

        $cl->reg_cb(
            session_ready => sub {
                my ( $cl, $acc ) = @_;
                if ($cnf) {
                    $muc->join_room( $acc->connection, $to, 'Giovanni' );
                    $muc_msg->{connection} = $acc->connection,
                    $muc_msg->{room} = $muc->get_room( $acc->connection, $to );
                    $muc_msg->send();
                }
                else {
                    $cl->send_message(
                        $msg => $to,
                        undef, 'chat'
                    );
                }
            },
            presence_update => sub {
                my ( $cl, $acc ) = @_;
                $cl->disconnect;
            },
            disconnect => sub {
                my ( $cl, $acc, $h, $p, $reas ) = @_;
                $j->broadcast;
            },
            error => sub {
                my ( $cl, $acc, $err ) = @_;
                print "ERROR: " . $err->string . "\n";
            },
        );
        $cl->start;
    }
    $j->wait;
};

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Giovanni::Plugins::Jabber

=head1 VERSION

version 0.9

=head1 AUTHOR

Lenz Gschwendtner <mail@norbu09.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by ideegeo Group Limited.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Giovanni/Plugins/HipChat.pm
package Giovanni::Plugins::HipChat;

use Mouse::Role;
use Data::Dumper;
use LWP::UserAgent;

around 'send_notify' => sub {
    my ( $orig, $self, $ssh ) = @_;

    print "notify via jabber\n";
    my @tos = split(/\s*,\s*/, $self->config->{hipchat_rooms});
    my $msg =
        'just ran a '
      . $self->config->{command} . ' for '
      . $self->config->{project} . ' on '
      . $ssh->get_host;
    my $ua = LWP::UserAgent->new();
    my $url = 'https://api.hipchat.com/v1/rooms/message?format=json&auth_token='.$self->config->{hipchat_token};
    foreach my $to (@tos){
        $ua->request(POST $url, {
            room_id => $to,
            from => 'Giovanni',
            message => $msg,
            message_format => 'text',
            notify => 1,
            color => 'green',
        });
    }
};

1;


lib/Giovanni/Plugins/Hipchat.pm
package Giovanni::Plugins::Hipchat;

use Mouse::Role;
use Data::Dumper;
use LWP::UserAgent;
use LWP::Protocol::https;

around 'send_notify' => sub {
    my ( $orig, $self, $ssh ) = @_;

    print "notify via HipChat\n";
    my @tos = split(/\s*,\s*/, $self->config->{hipchat_rooms});
    my $msg = $ENV{USER}
      . ' just ran a '
      . $self->config->{command} . ' for '
      . $self->config->{project} . ' on '
      . $ssh->get_host;
    my $ua = LWP::UserAgent->new();
    my $url = 'https://api.hipchat.com/v1/rooms/message?format=json&auth_token='.$self->config->{hipchat_token};
    foreach my $to (@tos){
        $ua->post($url, {
            room_id => $to,
            from => 'Giovanni',
            message => $msg,
            message_format => 'text',
            notify => 1,
            color => 'green',
        });
    }
};

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Giovanni::Plugins::Hipchat

=head1 VERSION

version 1.11

=head1 AUTHOR

Lenz Gschwendtner <mail@norbu09.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by ideegeo Group Limited.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Giovanni.pm
package Giovanni;

use 5.010;
use Mouse;
use Mouse::Util;
use Net::OpenSSH;
use Sys::Hostname;
use Cwd;
use Giovanni::Stages;

extends 'Giovanni::Stages';


our $VERSION = '1.9';

has 'debug' => (
    is        => 'rw',
    isa       => 'Bool',
    required  => 1,
    default   => 0,
    predicate => 'is_debug',
);

has 'hostname' => (
    is      => 'rw',
    isa     => 'Str',
    default => hostname(),
);

has 'repo' => (
    is       => 'rw',
    isa      => 'Str',
    required => 1,
    default  => cwd(),
);

has 'scm' => (
    is      => 'rw',
    isa     => 'Str',
    default => 'git',
);

has 'deploy_to' => (
    is      => 'rw',
    isa     => 'Str',
    default => '/var/www',
);

has 'user' => (
    is      => 'rw',
    isa     => 'Str',
    default => 'deploy',
);

has 'version' => (
    is      => 'rw',
    isa     => 'Str',
    default => 'v1',
);

has 'error' => (
    is  => 'rw',
    isa => 'Str',
);

has 'config' => (
    is       => 'rw',
    required => 1,
);

has 'notifyer' => (
    is      => 'rw',
    isa     => 'Str',
    default => 'jabber',
);


sub deploy {
    my ($self) = @_;

    # load SCM plugin
    $self->load_plugin($self->scm);
    my $tag = $self->tag();
    my $ssh = $self->_get_ssh_conn;
    $self->process_stages($ssh, 'deploy');
}

sub rollback {
    my ($self, $offset) = @_;

    my $ssh = $self->_get_ssh_conn;
    $self->process_stages($ssh, 'rollback');
}

sub process_stages {
    my ($self, $ssh, $mode) = @_;

    my $conf_key = ($mode eq 'restart' ? 'deploy' : $mode);
    my @stages = split(/\s*,\s*/, $self->config->{$conf_key});
    foreach my $stage (@stages) {
        if($mode eq 'restart'){
            next unless $stage =~ m/restart/;
            $self->process_hosts($ssh, $stage, $mode);

        } else {
            $self->process_hosts($ssh, $stage, $mode);

            # if one host produced an error while restarting, rollback all
            if ($self->error and ($stage =~ m/^restart/i) and ($mode eq 'deploy')) {
                $self->log('ERROR', $self->error);
                $self->process_stages($ssh, 'rollback');
                return;
            }
        }
    }
}

sub process_hosts {
    my ($self, $ssh, $stage, $mode) = @_;

    my @hosts = split(/\s*,\s*/, $self->config->{hosts});
    foreach my $host (@hosts) {
        $self->log($ssh->{$host}, "running $stage");
        $self->$stage($ssh->{$host});
    }
}

sub _get_ssh_conn {
    my ($self) = @_;

    my @hosts = split(/\s*,\s*/, $self->config->{hosts});
    my $ssh;
    foreach my $host (@hosts) {
        my $conn = $host;
        $conn = ($self->config->{user} || $self->user) . '@' . $host;
        $ssh->{$host} = Net::OpenSSH->new($conn, async => 1);
    }

    # trigger noop command to check for connection
    foreach my $host (@hosts) {
        $ssh->{$host}->test('echo')
            or confess "could not connect to $host: " . $ssh->{$host}->error;
        $self->log($host, 'connected');
    }
    return $ssh;
}

sub restart {
    my ($self) = @_;

    my $ssh = $self->_get_ssh_conn;
    $self->process_stages($ssh, 'restart');
}

sub load_plugin {
    my ($self, $plugin) = @_;

    my $plug = 'Giovanni::Plugins::' . ucfirst(lc($plugin));
    unless (Mouse::Util::is_class_loaded($plug)) {
        print STDERR "Loading $plugin Plugin\n" if $self->is_debug;
        with($plug);    # or confess "Could not load Plugin: '$plugin'\n";
    }

    return;
}

sub log {
    my ($self, $host, $log) = @_;

    return unless $log;

    my $name;
    given ($host) {
        when (ref $host eq 'Net::OpenSSH') { $name = $host->get_host; }
        default { $name = $host; }
    }
    chomp($log);
    print STDERR "[" . $name . "] " . $log . $/;

    return;
}

sub notify {
    my ($self, $ssh, $conf) = @_;

    # load notify plugin
    $self->load_plugin($self->notifyer);
    $self->send_notify($ssh, $conf);
    return;
}


1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Giovanni

=head1 VERSION

version 1.11

=head1 SYNOPSIS

Giovanni is a Perl replacement for the idea behind Capistrano. It is a
deployment system that can be used comfortably from the commandline to
check out code, restart systems and notify.
The system is currently used with git and manages some Catalyst and
Mojolicious apps and notifies via Jabber. It supports timestamped
rollouts (ie have the last 5 versions of your code on the server and link
to the currently running one) and plain git repositories. It tries to
detect problems in the deployment process and rolls back. It supports
manual rollbacks, two restart modes and does all that without any code
on the server. All you need is a working ssh setup with ssh-keys that
handle the login. 
We also use it with Jenkins to automatically deploy
code that successfully completed the test suite.

Giovanni comes with a commandline tool called gio. Check the gio manpage
for the config file format.

=head1 NAME

Giovanni - a Perl based deployment system

=head1 VERSION

Version 1.9

=head1 AUTHOR

Lenz Gschwendtner, C<< <norbu09 at cpan.org> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-giovanni at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Giovanni>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Giovanni

You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Giovanni>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Giovanni>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Giovanni>

=item * Search CPAN

L<http://search.cpan.org/dist/Giovanni/>

=back

=head1 ACKNOWLEDGEMENTS

=head1 LICENSE AND COPYRIGHT

Copyright 2011 Lenz Gschwendtner.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.

=head1 AUTHOR

Lenz Gschwendtner <mail@norbu09.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by ideegeo Group Limited.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut
